### PR TITLE
Build the test (assignment 2)

### PR DESCRIPTION
**Pull request by WentaoZhan1998**
This pull request is a part of assignment 2.

### CODE REVIEW

File: R/hiper_glm.R [hunk 1]
```diff
@@ -12,22 +28,32 @@
 #' @return A list including the information of the fitted model
```
Comment (chinandrew) [inline R/hiper_glm.R:12]: you'll need to update your documentation since you added an argument

File: R/hiper_glm.R [hunk 2]
```diff
@@ -41,13 +67,12 @@ hiper_glm <- function(design, outcome, model = "linear") {
...
     warning("`hiper_glm` for logit is yet to be implemented.")
   }
-  warning("`hiper_glm` is yet to be implemented.")
   hglm_out <- list()
   hglm_out$model = model
   hglm_out$coefficients = beta
   hglm_out$residuals = as.vector(res)
   hglm_out$fitted.values = as.vector(offsets)
-  hglm_out$XtX_inv = XtX_inv
+  #hglm_out$XtX_inv = XtX_inv
```
Comment (chinandrew) [inline R/hiper_glm.R]: -1

Don't leave unnecessary comments in. If you need to get old code, use your git history.

File: R/model_output.R [hunk 3]
```diff
@@ -6,8 +6,8 @@ print.hglm <- function(hglm_out) {
 
 #' @export
 coef.hglm <- function(hglm_out) {
-  sprintf('Model: ', hglm_out$model)
-  sprintf(hglm_out$coefficients)
+  sprintf(paste0('Model: ', hglm_out$model))
+  print(hglm_out$coefficients)
```
Comment (chinandrew) [inline R/model_output.R]: Keep things simple and just have `coef` return `hglm_out$coefficients`. No printing needed

File: tests/testthat/test-are-all-close.R [hunk 4]
```diff
@@ -0,0 +1,23 @@
+test_that("Vectors are all close", {
+  x = rep(1, 100)
```
Comment (chinandrew) [inline tests/testthat/test-are-all-close.R]: You don't really need 100 length vectors, but this is fine.

File: R/hiper_glm.R [hunk 5]
```diff
@@ -12,22 +28,32 @@
 #' @return A list including the information of the fitted model
 #'
 #' @export
-hiper_glm <- function(design, outcome, model = "linear") {
+hiper_glm <- function(design, outcome, model = "linear", option = NULL) {
   supported_model <- c("linear", "logit")
```
Comment (chinandrew) [inline R/hiper_glm.R]:
```suggestion
  supported_model <- c("linear")
```

You can also try using something like `match.arg()` here

File: R/hiper_glm.R [hunk 6]
```diff
@@ -12,22 +28,32 @@
...
   if (!(model %in% supported_model)) {
     stop(sprintf("The model %s is not supported.", model))
   }
   int = F
```
Comment (chinandrew) [inline R/hiper_glm.R]: -1

don't overengineer for now, keep things simple. And for design matrix validation I'd split this out into it's own function, see related comment below

File: R/hiper_glm.R [hunk 7]
```diff
@@ -12,22 +28,32 @@
...
+    }
+    else if(option$mle_solver == 'BFGS'){
+     log_lkl = function(beta, noise_var = 1){
+       .5*crossprod(outcome - crossprod(t(design), beta))[1,1]/noise_var
+     }
+     grad = function(beta, noise_var = 1){
+       grad_linear(design, outcome, beta = beta, noise_var = noise_var)
+     }
+     beta = optim(rep(0, ncol(design)), log_lkl, grad)$par
+    }
```
Comment (chinandrew) [inline R/hiper_glm.R]: -5

Break the solves out into their own functions, e.g.

```
 if(option["mle_solver"] == "BFGS"){
    model_coefs = mle_bfgs(...)
    } else {
    model_coefs = mle_pinv(...)
    }
```

Functions should be simple and only do one thing. In this case hiper_glm is just orchestrating the right function to call, and if you change how pinv works (e.g. going from LU to cholesky), hiper_glm never has to know (and shouldn't have to know)

Similarly, break out log likelihood into it's own function like you did with gradient. See comment on line 9 of test-mle-finder.R for related note on this.

File: R/hiper_glm.R [hunk 8]
```diff
@@ -41,13 +67,12 @@ hiper_glm <- function(design, outcome, model = "linear") {
   if (model == 'logit'){
     warning("`hiper_glm` for logit is yet to be implemented.")
```
Comment (chinandrew) [inline R/hiper_glm.R]: -1

This is redundant with the supported_model logic on line 33

File: R/hiper_glm.R [hunk 9]
```diff
@@ -12,22 +28,32 @@
...
 #'
 #' @export
-hiper_glm <- function(design, outcome, model = "linear") {
+hiper_glm <- function(design, outcome, model = "linear", option = NULL) {
   supported_model <- c("linear", "logit")
   if (!(model %in% supported_model)) {
     stop(sprintf("The model %s is not supported.", model))
   }
   int = F
   if (model == 'linear'){
```
Comment (chinandrew) [inline R/hiper_glm.R]: Don't mix single and double quotes. Tidyverse style guide recommends double quotes. Same elsewhere in code.

File: tests/testthat/test-are-all-close.R [hunk 10]
```diff
@@ -0,0 +1,23 @@
+test_that("Vectors are all close", {
```
Comment (chinandrew) [inline tests/testthat/test-are-all-close.R]: try to have the filename match the function or file its testing, so are_all_close and not are-all-close

File: tests/testthat/test-mle-finder.R [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  expect_equal(as.vector(beta_theory), as.vector(beta_numerical))
+})
+
+test_that("linalg and optim least-sq coincide", {
+  via_linalg_out <- hiper_glm(design, outcome, model = 'linear')
+  via_bfgs_out <- hiper_glm(
+    design, outcome, model = 'linear', option = list(mle_solver = 'BFGS')
+  )
+  expect_true(are_all_close(
+    coef(via_linalg_out), coef(via_bfgs_out), abs_tol = 1e-2, rel_tol = 1e-2
```
Comment (chinandrew) [inline tests/testthat/test-mle-finder.R]: This tolerances seem a little loose, you should be able to bring these down if you wanted

File: tests/testthat/test-mle-finder.R [hunk 12]
```diff
@@ -0,0 +1,24 @@
+n_obs <- 32; n_pred <- 4
+data <- simulate_data(n_obs, n_pred, model = 'linear', seed = 1918)
+design <- data$design; outcome <- data$outcome
+set.seed(1)
+beta = rnorm(ncol(design))
```
Comment (chinandrew) [inline tests/testthat/test-mle-finder.R]: -1 

Don't use randomness if you don't have to. Specifying a hardcoded `beta` works fine here and then the values are easy to know/read

If you do want to use it, e.g. to try a bunch of configurations, always use a seed as you have done.

File: R/hiper_glm.R [hunk 13]
```diff
@@ -1,3 +1,19 @@
+#' Gradient for the likelihood under linear model
+#'
+#' Calculate the gradient
+#'
+#' @details
+#'
+#' @param design matrix, the design matrix X for the predictors.
+#' @param outcome vector, the output Y for the response.
+#' @param beta vector, the potent8ial
```
Comment (chinandrew) [inline R/hiper_glm.R]:
```suggestion
#' @param beta vector, the potential
```

File: R/hiper_glm.R [hunk 14]
```diff
@@ -12,22 +28,32 @@
...
-      design = cbind(rep(1, nrow(design)), design)
+      #design = cbind(rep(1, nrow(design)), design)
       int = T
     }
     if(sum(apply(design,2, var)<=1e-10)>1){
       stop(sprintf("Too many intercept rows"))
     }
-    XtX_inv = solve(crossprod(design))
-    beta = XtX_inv %*% crossprod(design, outcome)
+    if(is.null(option$mle_solver)){
```
Comment (chinandrew) [inline R/hiper_glm.R]: If you're going to let people specify BFGS and the default is pseudo inv, I'd recommend also letting them specify pseudo inv explicitely. It's a bit weird if the default is not an explicitely specifiable option.

File: R/hiper_glm.R [hunk 15]
```diff
@@ -12,22 +28,32 @@
...
+      beta = solve(crossprod(design), crossprod(design, outcome))
+    }
+    else if(option$mle_solver == 'BFGS'){
```
Comment (chinandrew) [inline R/hiper_glm.R]: -1 

missing an `else` after the `else if`

File: tests/testthat/test-mle-finder.R [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+test_that("Gradients least-sq coincide", {
+  beta_theory = grad_linear(design, outcome, beta, noise_var = 1)
+  log_lkl = function(beta, noise_var = 1){
```
Comment (chinandrew) [inline tests/testthat/test-mle-finder.R]: As mentioned above, having log_lkl broken out would prevent you from having to write this twice (and if you change it in the main body of the code, you no longer are testing it properly here if you forgot to also change it here.

File: R/hiper_glm.R [hunk 17]
```diff
@@ -12,22 +28,32 @@
...
+     log_lkl = function(beta, noise_var = 1){
+       .5*crossprod(outcome - crossprod(t(design), beta))[1,1]/noise_var
```
Comment (chinandrew) [inline R/hiper_glm.R]: Be consistent with spacing between operators, e.g. `1 + 2 * 4` and not `1+2 * 4`. You can use an auto-formatter do handle this for you.

**Review by chinandrew — CHANGES_REQUESTED** [review]
-0 For EX1.
-0 For EX2A.
-0 For EX2B, though was confused since it wasnt in the Rmd file.

File: R/hiper_glm.R (hunk 2 above)
Comment (WentaoZhan1998) [reply R/hiper_glm.R]: Hi Andrew, I missed a typo in the pseudo-inverse part when I merged this pull request. So I made a follow-up one and you can go to here (https://github.com/WentaoZhan1998/hiperglm/pull/2) to see the changes.

Comment (chinandrew) [reply R/hiper_glm.R]: Thanks for the note, you can go ahead and merge that one too.

File: R/hiper_glm.R [hunk 18]
```diff
@@ -6,49 +6,34 @@
...
   if (!(model %in% supported_model)) {
     stop(sprintf("The model %s is not supported.", model))
   }
-  int = F
-  if (model == 'linear'){
-    if(sum(apply(design,2, var)<=1e-10)==0){
-      design = cbind(rep(1, nrow(design)), design)
-      int = T
+  if (model == "linear") {
+    if (option$mle_solver == "pinv") {
```
Comment (chinandrew) [inline R/hiper_glm.R:21]: if you wanted to add a default  option, you could do add a `|is.null(option$mlesolver )` to this condition

Comment (chinandrew) [reply R/hiper_glm.R:21]: Similarly you may want to add error handling if a non-valid solver option is supplied

**Comment by chinandrew** [comment]
Final grade 100/100

-0 For EX1.
-0 For EX2A.
-0 For EX2B, though was confused since it wasnt in the Rmd file.
-0 For EX3.

**Pull request merged**